src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -0,0 +1,440 @@
+    if (value != null) {
+      if (type.isPrimitive()) {
+        PrimitiveType primitiveType = type.asPrimitiveType();
+        switch (type.asPrimitiveType().getPrimitiveTypeName()) {
+          case BOOLEAN:
+            int intValue =
+                DataValidationUtil.validateAndParseBoolean(
+                    type.getName(), value, insertRowsCurrIndex);
+            value = intValue > 0;
+            stats.addIntValue(BigInteger.valueOf(intValue));
+            estimatedParquetSize += ParquetBufferValue.BIT_ENCODING_BYTE_LEN;
+            break;
+          case INT32:
+            int intVal = getInt32Value(value, primitiveType, insertRowsCurrIndex);
+            value = intVal;
+            stats.addIntValue(BigInteger.valueOf(intVal));
+            estimatedParquetSize += 4;
+            break;
+          case INT64:
+            long longVal =
+                getInt64Value(value, primitiveType, defaultTimezone, insertRowsCurrIndex);
+            value = longVal;
+            stats.addIntValue(BigInteger.valueOf(longVal));
+            estimatedParquetSize += 8;
+            break;
+          case FLOAT:
+            float floatVal =
+                DataValidationUtil.validateAndParseIcebergNumber(
+                    type.getName(), value, insertRowsCurrIndex, Float.class);
+            value = floatVal;
+            stats.addRealValue((double) floatVal);
+            estimatedParquetSize += 4;
+            break;
+          case DOUBLE:
+            double doubleVal =
+                DataValidationUtil.validateAndParseIcebergNumber(
+                    type.getName(), value, insertRowsCurrIndex, Double.class);
+            value = doubleVal;
+            stats.addRealValue(doubleVal);
+            estimatedParquetSize += 8;
+            break;
+          case BINARY:
+            byte[] byteVal = getBinaryValue(value, primitiveType, stats, insertRowsCurrIndex);
+            value = byteVal;
+            estimatedParquetSize +=
+                ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + byteVal.length;
+            break;
+          case FIXED_LEN_BYTE_ARRAY:
+            byte[] fixedLenByteArrayVal =
+                getFixedLenByteArrayValue(value, primitiveType, stats, insertRowsCurrIndex);
+            value = fixedLenByteArrayVal;
+            estimatedParquetSize +=
+                ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN
+                    + fixedLenByteArrayVal.length;
+            break;
+          default:
+            throw new SFException(
+                ErrorCode.UNKNOWN_DATA_TYPE,
+                type.getLogicalTypeAnnotation(),
+                primitiveType.getPrimitiveTypeName());
+        }
+      } else {
+        return getGroupValue(
+            value, type.asGroupType(), stats, defaultTimezone, insertRowsCurrIndex);
+      }
+    }
+
+    if (value == null) {
+      if (!type.isRepetition(Repetition.REQUIRED)) {
```

Comment:
what about primitive types, that are non-nullable, and value is still null?

Reply:
The logic was incorrect, fixed.